### PR TITLE
Akismet: Always cache key result

### PR DESCRIPTION
**Pull request by kraftbj**
#### Changes proposed in this Pull Request:

* Always cache the result of the Akismet key check.

#### Testing instructions:
* Do something that requires the Akismet check. For example, add a Email sharing button via the Sharing module to your site.
* Monitor your server's outbound HTTP and see that Akismet is checked no more than once on the front-end when looking at a post with an e-mail sharing button added.

#### Proposed changelog entry for your changes:

* Improves caching of Akismet status.

### CODE REVIEW

**Comment by mdawaffe** [comment]
I can see why we'd want to cache a failure for blog-side content. Do you think we should continue to *not* cache failure in wp-admin/ (and REST API?)?

**Comment by kraftbj** [comment]
I'd be open to re-checking if the cached value is false when in wp-admin. I don't see the harm in caching a false result when we get it, but be more open to checking anew in contexts where it'd be okay.

One thing to keep in mind is if the issue is a slow response from Akismet due to network traffic or something like that, it would possibly slow down various page loads or result in multiple requests being made at once on a site with many active users all within wp-admin. Less likely than front-end. 

In short, I would lean on keeping the cache but open to providing argument to force a recheck or something similar.

**Comment by mdawaffe** [comment]
(I haven't really looked at the code in any depth; please let me know if I'm wrong.)

If we cache a `failed` value, that would mean we can't check again for 10 minutes. If this function is used anywhere in wp-admin/ for displaying/configuring settings and if there's some odd one-off issue with the network connection, I'll get an error/won't see the settings I want/whatever, and I won't be able to do anything about it for 10 minutes even if the network issue resolved itself in 5 seconds.

I agree we shouldn't be making multiple HTTP requests to render one page, but we can limit things to one HTTP request per page load without persistent caching.

**Comment by kraftbj** [comment]
That's fair. I'll update it to not cache for the REST API or wp-admin requests. If we do end up with an edge case, we can look into it when that happens.

**Comment by mdawaffe** [comment]
The goal here is to never make multiple requests to Akismet on one page load. Is that correct?

With no caching at all in `is_admin()`, it seems like we'll run into another edge case like the original one you're protecting against with this patch. Should there be an in-memory cache (say, a `static` variable) on top of the transient cache?

**Comment by kraftbj** [comment]
The initial look I was taking was a high-traffic site having a remote call on every load, but you're right that an additional situation is multiple calls per page load.

Added a static variable and reorganized the function a bit.

**Review by jeherve — DISMISSED** [review]
This works well for me. @mdawaffe Want to give it another look and merge if that's okay?

Thank you!

File: class.jetpack.php [hunk 1]
```diff
@@ -7135,27 +7135,39 @@ function jetpack_user_col_style() {
...
+
+		// Do not used the cache result in wp-admin or REST API requests if the key isn't valid, in case someone is actively renewing, etc.
+		$recheck = ( ( is_admin() || ( defined( 'REST_REQUEST' ) && REST_REQUEST ) ) && 'valid' !== $akismet_key_state ) ? true : false;
+		// We cache the result of the Akismet key verification for ten minutes.
+		if ( ! $akismet_key_state || $recheck ) {
+			$akismet_key_state = Akismet::verify_key( $akismet_key );
+			set_transient( 'jetpack_akismet_key_is_valid', $akismet_key_state, 10 * MINUTE_IN_SECONDS );
+		}
+
+		$status = ( 'valid' === $akismet_key_state );
```
Comment (mdawaffe) [inline class.jetpack.php]: Nit: We don't need the parentheses.
```suggestion
		$status = 'valid' === $akismet_key_state;
```

**Review by mdawaffe — DISMISSED** [review]
I left some nits, but good to go as is.

**Review by jeherve — APPROVED** [review]
This should be good to merge now.

**Comment by jeherve** [comment]
Cherry-picked to `branch-6.8` in d6640b05fb659af9172222f4570f5784ac58f2f2
